interface/ceed-basis.c

```diff
@@ -623,6 +627,79 @@ int CeedBasisCreateH1(Ceed ceed, CeedElemTopology topo, CeedInt num_comp,
+                     dimension of which is used in some array sizes below
+  @param num_comp    Number of components (usually 1 for vectors in H(div) bases)
+  @param num_nodes   Total number of nodes (dofs per element)
+  @param num_qpts    Total number of quadrature points
+  @param interp      Row-major (dim*num_qpts * num_nodes) matrix expressing the values of
+                       nodal basis functions at quadrature points
+  @param div        Row-major (num_qpts * num_nodes) matrix expressing
+                       divergence of nodal basis functions at quadrature points
+  @param q_ref       Array of length num_qpts holding the locations of quadrature
+                       points on the reference element
+  @param q_weight    Array of length num_qpts holding the quadrature weights on the
+                       reference element
+  @param[out] basis  Address of the variable where the newly created
+                       CeedBasis will be stored.
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref User
+**/
+int CeedBasisCreateHdiv(Ceed ceed, CeedElemTopology topo, CeedInt num_comp,
+                        CeedInt num_nodes, CeedInt num_qpts, const CeedScalar *interp,
+                        const CeedScalar *div, const CeedScalar *q_ref,
+                        const CeedScalar *q_weight, CeedBasis *basis) {
+  int ierr;
+  CeedInt Q = num_qpts, P = num_nodes, dim = 0;
+  ierr = CeedBasisGetTopologyDimension(topo, &dim); CeedChk(ierr);
+  if (!ceed->BasisCreateHdiv) {
+    Ceed delegate;
+    ierr = CeedGetObjectDelegate(ceed, &delegate, "Basis"); CeedChk(ierr);
+
+    if (!delegate)
+      // LCOV_EXCL_START
+      return CeedError(ceed, CEED_ERROR_UNSUPPORTED,
+                       "Backend does not implement BasisCreateHdiv");
+    // LCOV_EXCL_STOP
+
+    ierr = CeedBasisCreateHdiv(delegate, topo, num_comp, num_nodes,
+                               num_qpts, interp, div, q_ref,
+                               q_weight, basis); CeedChk(ierr);
+    return CEED_ERROR_SUCCESS;
+  }
+
+  ierr = CeedCalloc(1,basis); CeedChk(ierr);
+
+  (*basis)->ceed = ceed;
+  ierr = CeedReference(ceed); CeedChk(ierr);
+  (*basis)->ref_count = 1;
+  (*basis)->tensor_basis = 0;
+  (*basis)->dim = dim;
+  (*basis)->topo = topo;
+  (*basis)->num_comp = num_comp;
+  (*basis)->P = P;
+  (*basis)->Q = Q;
+  (*basis)->Q_comp = dim;
+  (*basis)->basis_space = 2; // 2 for H(div) space
+  ierr = CeedMalloc(Q*dim,&(*basis)->q_ref_1d); CeedChk(ierr);
+  ierr = CeedMalloc(Q,&(*basis)->q_weight_1d); CeedChk(ierr);
+  if (q_ref) memcpy((*basis)->q_ref_1d, q_ref, Q*dim*sizeof(q_ref[0]));
+  if(q_weight) memcpy((*basis)->q_weight_1d, q_weight, Q*sizeof(q_weight[0]));
+  ierr = CeedMalloc(dim*Q*P, &(*basis)->interp); CeedChk(ierr);
+  ierr = CeedMalloc(Q*P, &(*basis)->div); CeedChk(ierr);
+  if(interp) memcpy((*basis)->interp, interp, dim*Q*P*sizeof(interp[0]));
+  if(div) memcpy((*basis)->div, div, Q*P*sizeof(div[0]));
```

Comment:
minor style to reflect main
```suggestion
  ierr = CeedCalloc(1, basis); CeedChk(ierr);

  (*basis)->ceed = ceed;
  ierr = CeedReference(ceed); CeedChk(ierr);
  (*basis)->ref_count = 1;
  (*basis)->tensor_basis = 0;
  (*basis)->dim = dim;
  (*basis)->topo = topo;
  (*basis)->num_comp = num_comp;
  (*basis)->P = P;
  (*basis)->Q = Q;
  (*basis)->Q_comp = dim;
  (*basis)->basis_space = 2; // 2 for H(div) space
  ierr = CeedMalloc(Q*dim, &(*basis)->q_ref_1d); CeedChk(ierr);
  ierr = CeedMalloc(Q, &(*basis)->q_weight_1d); CeedChk(ierr);
  if (q_ref) memcpy((*basis)->q_ref_1d, q_ref, Q*dim*sizeof(q_ref[0]));
  if (q_weight) memcpy((*basis)->q_weight_1d, q_weight, Q*sizeof(q_weight[0]));
  ierr = CeedMalloc(dim*Q*P, &(*basis)->interp); CeedChk(ierr);
  ierr = CeedMalloc(Q*P, &(*basis)->div); CeedChk(ierr);
  if (interp) memcpy((*basis)->interp, interp, dim*Q*P*sizeof(interp[0]));
  if (div) memcpy((*basis)->div, div, Q*P*sizeof(div[0]));
```